source/extensions/filters/udp/udp_proxy/udp_proxy_filter.cc

```diff
@@ -0,0 +1,86 @@
+#include "extensions/filters/udp/udp_proxy/udp_proxy_filter.h"
+
+#include "envoy/network/listener.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace UdpFilters {
+namespace UdpProxy {
+
+// TODO(mattklein123): Logging
+// TODO(mattklein123): Stats
+
+void UdpProxyFilter::onData(Network::UdpRecvData& data) {
+  const auto active_session_it = sessions_.find(data.addresses_);
+  ActiveSession* active_session;
+  if (active_session_it == sessions_.end()) {
+    // TODO(mattklein123): Session circuit breaker.
+    // TODO(mattklein123): Instead of looking up the cluster each time, keep track of it via
+    // cluster manager callbacks.
+    Upstream::ThreadLocalCluster* cluster = config_->getCluster();
+    // TODO(mattklein123): Handle the case where the cluster does not exist.
+    ASSERT(cluster != nullptr);
+
+    // TODO(mattklein123): Pass a context and support hash based routing.
+    Upstream::HostConstSharedPtr host = cluster->loadBalancer().chooseHost(nullptr);
+    // TODO(mattklein123): Handle the case where the host does not exist.
+    ASSERT(host != nullptr);
+
+    auto new_session = std::make_unique<ActiveSession>(*this, std::move(data.addresses_), host);
+    active_session = new_session.get();
+    sessions_.emplace(std::move(new_session));
+  } else {
+    // TODO(mattklein123): Handle the host going away going away or failing health checks.
+    active_session = active_session_it->get();
+  }
+
+  active_session->write(*data.buffer_);
+}
+
+UdpProxyFilter::ActiveSession::ActiveSession(UdpProxyFilter& parent,
+                                             Network::UdpRecvData::LocalPeerAddresses&& addresses,
+                                             const Upstream::HostConstSharedPtr& host)
+    : parent_(parent), addresses_(std::move(addresses)), host_(host),
+      io_handle_(host->address()->socket(Network::Address::SocketType::Datagram)),
```

Comment:
Is it possible that the socket creating and binding fail here? This is getting a random port right? Can we run out of ports?

Reply:
I'm not sure. AFAIK, it will not actually fail until the first datagram is sent, as that is when an ephemeral port is acquired. I *think* this can only fail if we run out of FDs, and in that case we crash. Will add comments and look more.